lib/bandit/http1/handler.ex

```diff
@@ -88,6 +100,50 @@ defmodule Bandit.HTTP1.Handler do
     end
   end
 
+  defp maybe_upgrade_h2c(req, transport_info, method, request_target, headers) do
+    with {:upgrade, "h2c"} <- {:upgrade, Bandit.Headers.get_header(headers, "upgrade")},
+         %Bandit.TransportInfo{secure?: false} <- transport_info,
+         {:ok, connection_headers} <- get_connection_header_keys(headers),
+         {:settings, [{"http2-settings", settings_payload}]} <-
+           {:settings, Enum.filter(headers, fn {key, _value} -> key == "http2-settings" end)},
+         {:ok, settings_payload} <- Base.url_decode64(settings_payload) do
+      Bandit.HTTP1.Adapter.send_resp(
+        req,
+        101,
+        [{"connection", "Upgrade"}, {"upgrade", "h2c"}],
+        <<>>
+      )
+
+      headers =
+        Enum.reject(headers, fn {key, _value} ->
+          key == "connection" || key in connection_headers
+        end)
+
+      {:ok, :h2c, settings_payload, method, request_target, headers}
+    else
+      %Bandit.TransportInfo{secure?: true} -> {:error, "h2c is only supported on http"}
+      {:upgrade, _} -> {:ok, :no_upgrade}
+      {:settings, _} -> {:error, "Expected 1 http2-settings header"}
+      :error -> {:error, "Invalid http2-settings value"}
+    end
+  end
+
+  defp get_connection_header_keys(headers) do
+    case Bandit.Headers.get_header(headers, "connection") do
+      nil ->
+        {:error, "Expected connection header"}
+
+      value ->
+        header_keys =
+          value
+          |> String.downcase()
+          |> String.split(",")
+          |> Enum.map(&String.trim/1)
+
+        {:ok, header_keys}
+    end
+  end
```

Comment:
This probably belongs in Bandit.Header. I think there may be a helper for this in Plug.Utils as well?

Reply:
Great. Didn't know about the util function. Done in 5152ee0

(However note that `Plug.Conn.Utils.list/1` in plug version 1.14.2 does not trim trailing spaces but the main branch is now fixed)

Reply:
> (However note that `Plug.Conn.Utils.list/1` in plug version 1.14.2 does not trim trailing spaces but the main branch is now fixed)

There is a skipped test related to this as well. 

See:
https://github.com/mtrudel/bandit/blob/main/test/bandit/headers_test.exs#L69-L78